### PR TITLE
Try a new approach for caching Maven local repository

### PR DESCRIPTION
**Pull request by gsmet**
We used to cache a repo once a week but only after building the Java 17 build.

The new approach is the following:
- We use dependency:go-offline to populate the cache, hoping the quality will be better
- We drop the cache entirely every month
- We have one cache entry per branch and we keep updating it during the month by creating new caches
- We use restore-keys to find the best cache possible for a given run
- We share this cache with all the workflows

Let's create it as draft to check what's going on for the main workflow first. The secondary workflows can be fixed later.

### CODE REVIEW

**Comment by gsmet** [comment]
I think this is ready for review. There might be some issues in the non-PR workflows but I'll monitor things once this is in.

Let's wait for a full CI run of the PR workflow, though.

File: .github/workflows/ci-actions-incremental.yml [hunk 1]
```diff
@@ -129,17 +131,31 @@ jobs:
       - name: Generate cache key
         id: cache-key
         run: |
+          CURRENT_BRANCH="${{ github.repository != 'quarkusio/quarkus' && 'fork' || github.base_ref || github.ref_name }}"
+          CURRENT_MONTH=$(/bin/date -u "+%Y-%m")
+          CURRENT_DAY=$(/bin/date -u "+%d")
+          ROOT_CACHE_KEY="m2-cache"
+          echo "m2-monthly-cache-key=${ROOT_CACHE_KEY}-${CURRENT_MONTH}" >> $GITHUB_OUTPUT
+          echo "m2-monthly-branch-cache-key=${ROOT_CACHE_KEY}-${CURRENT_MONTH}-${CURRENT_BRANCH}" >> $GITHUB_OUTPUT
+          echo "m2-cache-key=${ROOT_CACHE_KEY}-${CURRENT_MONTH}-${CURRENT_BRANCH}-${CURRENT_DAY}" >> $GITHUB_OUTPUT
```
Comment (gastaldi) [inline .github/workflows/ci-actions-incremental.yml:140]: Wonder if we can create an action to do that for us

File: .github/workflows/ci-actions-incremental.yml [hunk 2]
```diff
@@ -129,17 +131,31 @@ jobs:
       - name: Generate cache key
         id: cache-key
         run: |
+          CURRENT_BRANCH="${{ github.repository != 'quarkusio/quarkus' && 'fork' || github.base_ref || github.ref_name }}"
```
Comment (gastaldi) [inline .github/workflows/ci-actions-incremental.yml:134]: What is `fork`?

**Review by gastaldi — APPROVED** [review]
LGTM, let's see how it goes

File: .github/workflows/ci-actions-incremental.yml (hunk 1 above)
Comment (gsmet) [reply .github/workflows/ci-actions-incremental.yml:140]: Yeah I had the same thought. I might do it at some point. For now, I think it's fine even if a bit repetitive.

Ideally I think we would even need to include the cache call into it.

File: .github/workflows/ci-actions-incremental.yml (hunk 2 above)
Comment (gsmet) [reply .github/workflows/ci-actions-incremental.yml:134]: So this is the equivalent of the ternary operator in GitHub world.

In the case of a fork, we don't want to use the branch as the name of the cache, as you use feature branch. So for forks, we are just using `fork` as the equivalent of the branch name.

File: .github/workflows/ci-actions-incremental.yml (hunk 1 above)
Comment (gsmet) [reply .github/workflows/ci-actions-incremental.yml:140]: Also Quarkus GitHub Actions are composite actions so we could even bundle the cache call into it.

**Pull request merged**